### ISSUE TITLE
Speed ​​up ResumeImpl

### ISSUE THREAD

**Issue by huangmengbin** [facebook/rocksdb#12160]
- We use rocksdb on a remote distributed file system. Sometimes, an IOError during `WAL` may cause the db instance to go into read-only mode，so we call `DB::Resume()` manually. However, `WaitForBackgroundWork` is required here, which may take a long time. Is it necessary in this case? Can we skip this step? Thanks !

**Comment by ajkr** [comment]
Yes, that sounds slow. Maybe we can use `PauseBackgroundWork()` when the error blocks foreground operations. I believe that is `Status::Severity::kHardError`

Do you mind clarifying what kind of `IOError` we are talking about? In particular I'm interested if you are setting the scope, subcode, or any other fields.

**Comment by huangmengbin** [comment]
- Some fields:
Reason=2 (kWriteCallback)
Code=5 (kIOError)
SubCode=0
Retryable=true

- Afterwards, it will be set to `kHardError` in `ErrorHandler::SetBGError`

- We found that `bg_compaction_paused_` is only used in `MaybeScheduleFlushOrCompaction`. It seems that tasks already scheduled will not stop immediately.

**Comment by huangmengbin** [comment]
Thanks for your reply ! We want to fix this kind of error by flushing manually:
 1. Reject all write requests for this db
 2. Override the bg_err status in `OnBackgroundError`
 3. Call `DB::Flush`

Will it be ok?

**Comment by ajkr** [comment]
> Thanks for your reply !
> 
> * Some fields:
>   Reason=2 (kWriteCallback)
>   Code=5 (kIOError)
>   SubCode=0
>   Retryable=true
> * In `ErrorHandler::SetBGError`, it will be set to `kHardError`
> * We found that `bg_compaction_paused_` is only used in `MaybeScheduleFlushOrCompaction`. It seems that tasks already scheduled will not stop immediately.

Oh I see, you are setting retryable to true and leaving scope at its default. That's the same as what we do. Thanks for the details.

> We found that bg_compaction_paused_ is only used in MaybeScheduleFlushOrCompaction. It seems that tasks already scheduled will not stop immediately.

Nice catch, I got confused by an inconsistency. Pausing manual compaction causes early termination of running manual compactions: https://github.com/facebook/rocksdb/blob/4fefe1fed9092be69d9771f2df7cdfadcf604bc3/db/compaction/compaction_iterator.cc#L460-L461

**Comment by ajkr** [comment]
> We want to fix this kind of error by flushing manually:

Yeah, you can probably work around it.

Regarding `ErrorHandler`'s bottleneck, I was never really convinced that we need to distinguish errors during recovery from errors outside of recovery. If we do not really need this distinction, then the `WaitForBackgroundWork()` can eventually be removed. It might take a bit of time to rework that code though.
